sql: explain crash when showing spans for virtual table

This was discovered while attempting to run logic tests with statement diagnostics on all the time.

```
CREATE TABLE ab(a INT, b INT);
EXPLAIN (VERBOSE) SELECT attnum FROM pg_attribute WHERE attrelid = 'ab'::regclass AND attname = 'b';
*
* ERROR: [n1,client=127.0.0.1:34888,hostssl,user=demo] a SQL panic has occurred while executing the following statement:
* EXPLAIN (VERBOSE) SELECT attnum FROM pg_attribute WHERE (attrelid = 'ab'::REGCLASS) AND (attname = 'b')
*
*
* ERROR: [n1,client=127.0.0.1:34888,hostssl,user=demo] a panic has occurred!
* panic: column family 0 not found
* (1) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:479
*   | [...repeated from below...]
* Wraps: (2) while executing: EXPLAIN (VERBOSE) SELECT _ FROM _ WHERE (_ = _::REGCLASS) AND (_ = _)
* Wraps: (3) attached stack trace
*   -- stack trace:
*   | github.com/cockroachdb/cockroach/pkg/sql.(*Server).ServeConn.func1
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:479
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:969
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror.CatchVectorizedRuntimeError.func1
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror/error.go:60
*   | runtime.gopanic
*   | 	/usr/local/go/src/runtime/panic.go:969
*   | github.com/cockroachdb/cockroach/pkg/sql/rowenc.NeededColumnFamilyIDs
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowenc/index_encoding.go:361
*   | github.com/cockroachdb/cockroach/pkg/sql/span.(*Builder).appendSpansFromConstraintSpan
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/span/span_builder.go:275
*   | github.com/cockroachdb/cockroach/pkg/sql/span.(*Builder).SpansFromConstraint
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/span/span_builder.go:228
*   | github.com/cockroachdb/cockroach/pkg/sql.generateScanSpans
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt_exec_factory.go:146
*   | github.com/cockroachdb/cockroach/pkg/sql.emitExplain.func1
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:164
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.(*emitter).spansStr
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:728
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.(*emitter).emitSpans
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:715
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.(*emitter).emitNodeAttributes
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:379
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit.func1
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:51
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit.func1
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:55
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit.func1
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:55
*   | github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain.Emit
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain/emit.go:64
*   | github.com/cockroachdb/cockroach/pkg/sql.emitExplain
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:183
*   | github.com/cockroachdb/cockroach/pkg/sql.(*explainPlanNode).startExec
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/explain_plan.go:127
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec.func2
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:504
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal.func1
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:119
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visitInternal
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:286
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planVisitor).visit
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:86
*   | github.com/cockroachdb/cockroach/pkg/sql.walkPlan
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/walk.go:50
*   | github.com/cockroachdb/cockroach/pkg/sql.startExec
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan.go:507
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planNodeToRowSource).Start
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/plan_node_to_row_source.go:126
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).Init
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:133
*   | github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror.CatchVectorizedRuntimeError
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexecbase/colexecerror/error.go:93
*   | github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Materializer).Start
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/materializer.go:242
*   | github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).Run
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:764
*   | github.com/cockroachdb/cockroach/pkg/sql/flowinfra.(*FlowBase).Run
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/flowinfra/flow.go:380
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).Run
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:366
*   | github.com/cockroachdb/cockroach/pkg/sql.(*DistSQLPlanner).PlanAndRun
*   | 	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsql_running.go:959
* Wraps: (4) panic: column family 0 not found
* Error types: (1) *withstack.withStack (2) *safedetails.withSafeDetails (3) *withstack.withStack (4) *errutil.leafError
*
panic: column family 0 not found [recovered]
	panic: column family 0 not found [recovered]
	panic: panic: column family 0 not found
```

--- a/pkg/sql/explain_plan.go
+++ b/pkg/sql/explain_plan.go
@@ -18,8 +18,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/keys"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catalogkeys"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
-	"github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc"
 	"github.com/cockroachdb/cockroach/pkg/sql/colflow"
 	"github.com/cockroachdb/cockroach/pkg/sql/execinfrapb"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/cat"
@@ -27,6 +25,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/exec/explain"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
 	"github.com/cockroachdb/cockroach/pkg/sql/sessiondatapb"
+	"github.com/cockroachdb/cockroach/pkg/util/errorutil"
 	"github.com/cockroachdb/errors"
 )
 
@@ -150,17 +149,29 @@ func emitExplain(
 	evalCtx *tree.EvalContext,
 	codec keys.SQLCodec,
 	explainPlan *explain.Plan,
-) error {
-	spanFormatFn := func(table cat.Table, index cat.Index, scanParams exec.ScanParams) string {
-		var tabDesc *tabledesc.Immutable
-		var idxDesc *descpb.IndexDescriptor
-		if table.IsVirtualTable() {
-			tabDesc = table.(*optVirtualTable).desc
-			idxDesc = index.(*optVirtualIndex).desc
-		} else {
-			tabDesc = table.(*optTable).desc
-			idxDesc = index.(*optIndex).desc
+) (err error) {
+	// Guard against bugs in the explain code.
+	defer func() {
+		if r := recover(); r != nil {
+			// This code allows us to propagate internal and runtime errors without
+			// having to add error checks everywhere throughout the code. This is only
+			// possible because the code does not update shared state and does not
+			// manipulate locks.
+			if ok, e := errorutil.ShouldCatch(r); ok {
+				err = e
+			} else {
+				// Other panic objects can't be considered "safe" and thus are
+				// propagated as crashes that terminate the session.
+				panic(r)
+			}
 		}
+	}()
+	spanFormatFn := func(table cat.Table, index cat.Index, scanParams exec.ScanParams) string {
+		if table.IsVirtualTable() {
+			return "<virtual table spans>"
+		}
+		tabDesc := table.(*optTable).desc
+		idxDesc := index.(*optIndex).desc
 		spans, err := generateScanSpans(evalCtx, codec, tabDesc, idxDesc, scanParams)
 		if err != nil {
 			return err.Error()

--- a/pkg/sql/opt/exec/execbuilder/testdata/explain
+++ b/pkg/sql/opt/exec/execbuilder/testdata/explain
@@ -1431,3 +1431,29 @@ vectorized: true
                       missing stats
                       table: u@primary
                       spans: [/1 - /1]
+
+# Make sure EXPLAIN (VERBOSE) works when there is a constrained scan of a
+# virtual table (#58193).
+statement ok
+CREATE TABLE ab (a INT, b INT)
+
+query T
+EXPLAIN (VERBOSE) SELECT attnum FROM pg_attribute WHERE attrelid = 'ab'::regclass AND attname = 'b';
+----
+distribution: local
+vectorized: true
+·
+• project
+│ columns: (attnum)
+│ estimated row count: 1 (missing stats)
+│
+└── • filter
+    │ columns: (attrelid, attname, attnum)
+    │ estimated row count: 1 (missing stats)
+    │ filter: attname = 'b'
+    │
+    └── • virtual table
+          columns: (attrelid, attname, attnum)
+          estimated row count: 10 (missing stats)
+          table: pg_attribute@pg_attribute_attrelid_idx
+          spans: [/ab - /ab]

--- a/pkg/sql/opt/exec/execbuilder/testdata/scalar
+++ b/pkg/sql/opt/exec/execbuilder/testdata/scalar
@@ -1061,7 +1061,7 @@ vectorized: true
   columns: (attrelid, attname, atttypid, attstattarget, attlen, attnum, attndims, attcacheoff, atttypmod, attbyval, attstorage, attalign, attnotnull, atthasdef, attidentity, attgenerated, attisdropped, attislocal, attinhcount, attcollation, attacl, attoptions, attfdwoptions)
   estimated row count: 10 (missing stats)
   table: pg_attribute@pg_attribute_attrelid_idx
-  spans: /53-/54
+  spans: [/t - /t]
 
 query T
 EXPLAIN (VERBOSE) SELECT CASE WHEN current_database() = 'test' THEN 42 ELSE 1/3 END

--- a/pkg/sql/opt/exec/explain/emit.go
+++ b/pkg/sql/opt/exec/explain/emit.go
@@ -116,8 +116,9 @@ func Emit(plan *Plan, ob *OutputBuilder, spanFormatFn SpanFormatFn) error {
 	return nil
 }
 
-// SpanFormatFn is a function used to format spans for EXPLAIN. Only called when
-// there is an index constraint or an inverted constraint.
+// SpanFormatFn is a function used to format spans for EXPLAIN. Only called on
+// non-virtual tables, when there is an index constraint or an inverted
+// constraint.
 type SpanFormatFn func(table cat.Table, index cat.Index, scanParams exec.ScanParams) string
 
 // omitTrivialProjections returns the given node and its result columns and
@@ -723,8 +724,8 @@ func (e *emitter) spansStr(table cat.Table, index cat.Index, scanParams exec.Sca
 		return "FULL SCAN"
 	}
 
-	// In verbose mode show the physical spans.
-	if e.ob.flags.Verbose {
+	// In verbose mode show the physical spans, unless the table is virtual.
+	if e.ob.flags.Verbose && !table.IsVirtualTable() {
 		return e.spanFormatFn(table, index, scanParams)
 	}
 

--- a/pkg/sql/rowenc/index_encoding.go
+++ b/pkg/sql/rowenc/index_encoding.go
@@ -358,7 +358,7 @@ func NeededColumnFamilyIDs(
 		return nil
 	})
 	if family0 == nil {
-		panic("column family 0 not found")
+		panic(errors.AssertionFailedf("column family 0 not found"))
 	}
 
 	// If all the needed families are nullable, we also need family 0 as a
@@ -1367,11 +1367,11 @@ func EncodeSecondaryIndexes(
 	indexBoundAccount mon.BoundAccount,
 ) ([]IndexEntry, error) {
 	if len(secondaryIndexEntries) > 0 {
-		panic("Length of secondaryIndexEntries was non-zero")
+		panic(errors.AssertionFailedf("length of secondaryIndexEntries was non-zero"))
 	}
 
 	if indexBoundAccount.Monitor() == nil {
-		panic("Memory monitor passed to EncodeSecondaryIndexes was nil")
+		panic(errors.AssertionFailedf("memory monitor passed to EncodeSecondaryIndexes was nil"))
 	}
 	const sizeOfIndexEntry = int64(unsafe.Sizeof(IndexEntry{}))
 

